### ISSUE TITLE
Raspberry pi builds broken due to partition table choice

### ISSUE THREAD

**Issue by MoralCode** [osbuild/bootc-image-builder#701]
I'm trying to follow https://mrguitar.net/?p=2605 to revive my raspberry pi server with bootc (with minor modifications since i only have a raspberry pi 3 not a pi 4)

While troubleshooting, I discovered https://github.com/osbuild/bootc-image-builder/issues/551#issuecomment-2245120613, which seems to point to the partition table configs being based on those from fedora iot (where arm builds used dos partitions explicitly for raspi compatibility). It seems like this setting was changed in bib for arm builds so that everything could be gpt instead. 

I did some [further reading](https://raspberrypi.stackexchange.com/questions/39959/raspbian-boot-process-and-the-partition-table/39969#39969) and it seems like the weird gpu-first way that a raspi boots **depends** on having a dos format partition table as it looks for critical files in the "first msdos partition on the card"

**Comment by MoralCode** [comment]
Ping as requested @supakeen

**Comment by supakeen** [comment]
Thank you.

**Comment by ondrejbudai** [comment]
If I'm reading the log correctly, this is the last version of bib before we switched [aarch64 to GPT](https://github.com/osbuild/bootc-image-builder/pull/582): quay.io/centos-bootc/bootc-image-builder:sha256-465c82cf63b450408fc61f3962fe76423b6c9a3c5bf84b55353605e3604f8c4d

Alternatively, you can build one yourself from git.

@achilleas-k  is currently working [on new partitioning](https://issues.redhat.com/browse/COMPOSER-2301). While GPT vs. MBR is not in scope, it should be a simple addition, that I would absolutely be in favor of. :)

**Comment by supakeen** [comment]
The alternative would be, as suggested by @mvo5 in https://github.com/osbuild/bootc-image-builder/pull/586 to introduce an explicit switch to use a 'Pi-compatible' partition table. We can use the same partition table as we use for Fedora IoT and Fedora Minimal on aarch64 as we know these work.

However, as @ondrejbudai pointed out in his experiments. `bootupd` does not copy the necessary firmware and dtb into `/boot` when installing the bootloader. The linked blogpost does this manually after-the-fact (by mounting the disk image and moving the appropriate files around) and in the experiment (https://github.com/ondrejbudai/fedora-bootc-raspi/tree/main) there's a hacky `bootupctl` way to do it at build time.

That's to say that if we provide an easy way to switch partition layout there are still things that need to be done manually to make the image bootable.

**Comment by MoralCode** [comment]
Poking through some of these links, it sounds like other SBC's also use a similar "read files from the root of the first MBR partition found" method of starting up, so maybe its worth generalizing this for all SBC's. Even if not all of them are "supported"/in there by default i think its important to avoid breaking compatibility with other SBC's, so that people would only need to add their SBC partition table definition and possibly specify what firmware to install/copy to the efi partition for it to work

**Comment by achilleas-k** [comment]
> @achilleas-k is currently working [on new partitioning](https://issues.redhat.com/browse/COMPOSER-2301). While GPT vs. MBR is not in scope, it should be a simple addition, that I would absolutely be in favor of. :)

I'll add it to config then.

**Comment by ondrejbudai** [comment]
> Poking through some of these links, it sounds like other SBC's also use a similar "read files from the root of the first MBR partition found" method of starting up, so maybe its worth generalizing this for all SBC's. Even if not all of them are "supported"/in there by default i think its important to avoid breaking compatibility with other SBC's, so that people would only need to add their SBC partition table definition and possibly specify what firmware to install/copy to the efi partition for it to work

Agreed! But in the current setup, I strongly believe that `bootupd` should be the one responsible for actually installing the firmware files. Implementing this in `bootc-image-builder` feels a bit out of the place. It would be nice, though, if we can somehow share the config between `bootupd` and `bootc-image-builder`. We have some ideas floating around about this.

**Comment by supakeen** [comment]
> > @achilleas-k is currently working [on new partitioning](https://issues.redhat.com/browse/COMPOSER-2301). While GPT vs. MBR is not in scope, it should be a simple addition, that I would absolutely be in favor of. :)
> 
> I'll add it to config then.

Since these have landed would this now be possible to document or implement (minus the firmware copying).

**Comment by MoralCode** [comment]
> However, as @ondrejbudai pointed out in his experiments. `bootupd` does not copy the necessary firmware and dtb into `/boot` when installing the bootloader.

After booting a current build of Raspi OS to my pi3-ish - it seems like theyve moved all the firmware files to `/boot/firmware`, so maybe this makes it easier for `bootupd` to do its thing?

**Comment by ondrejbudai** [comment]
Note that Fedora uses a different firmware and bootloader setup, so unless someone makes a bootc image of Raspi OS, I'm afraid this change is mostly irrelevant for us. :(

**Comment by MoralCode** [comment]
> Note that Fedora uses a different firmware and bootloader setup

Fedora in general or specifically the bootc versions?

> so unless someone makes a bootc image of Raspi OS, I'm afraid this change is mostly irrelevant for us. :(

F40 already has arm builds available that claim to support the raspi: https://docs.fedoraproject.org/en-US/quick-docs/raspberry-pi/ - would that help?

**Comment by ondrejbudai** [comment]
Fedora in general uses a different firmware and bootloader setup for raspi. But honestly I think it's not that important where the files are located, it's just that bootupd needs to learn how to put extra files anywhere. :)

**Comment by ondrejbudai** [comment]
ok, GPT vs. MBR support is here: https://github.com/osbuild/images/pull/1085

next thing is custom partition types (raspi3 needs 06 instead of EF for ESP)

**Comment by martin-paulus** [comment]
Would you welcome a set of pull requests which would introduce a Raspberry Pi-specific partition table type?

The blueprint snippet could look like this, for example:
```
{
  "customizations": {
    "disk": {
      "type": "rpi",
      "partitions": [
        {
          "fs_type": "vfat",
          "label": "ESP",
          "type": "plain",
          "minsize": "99 MiB",
          "mountpoint": "/boot/efi"
        }
      ]
    }
  }
}
```

And in images (pkg/disk/disk.go):
```
const (
	PT_NONE PartitionTableType = iota
	PT_DOS
	PT_GPT
	PT_RPI
)
```

```
func getPartitionTypeIDfor(ptType PartitionTableType, partTypeName string, architecture arch.Arch) (string, error) {
	switch ptType {
	case PT_RPI:
		switch partTypeName {
		case "data", "boot", "root", "usr":
			return FilesystemLinuxDOSID, nil
		case "esp":
			return FAT16BDOSID, nil
		case "lvm":
			return LVMPartitionDOSID, nil
		case "swap":
			return SwapPartitionDOSID, nil
		default:
			return "", fmt.Errorf("unknown or unsupported partition type name: %s", partTypeName)
		}
...
```
